pkg/kvstore/allocator/allocator.go

```diff
@@ -131,6 +131,18 @@ func (k *kvstoreBackend) AllocateID(ctx context.Context, id idpool.ID, key alloc
 	return nil
 }
 
+// AllocateID allocates a key->ID mapping in the kvstore.
+func (k *kvstoreBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	// create /id/<ID> and fail if it already exists
+	keyPath := path.Join(k.idPrefix, id.String())
+	success, err := kvstore.CreateOnlyIfLocked(ctx, keyPath, []byte(key.GetKey()), false, lock)
+	if err != nil || !success {
```

Comment:
Maybe `switch` and have a different error returned depending on if the error is non-nil or `!success` since if the error is nil, but `success == false` we will return a nil string in the error we create.

Reply:
I did start down that road but the original code conflates `err != nil` and `!success` so I left it as-is.  Looking at the pkg/kvstore/etcd.go  implementation it seems like `!success` is always `err != nil` (there are hardcoded `return false, err` statements).
https://github.com/cilium/cilium/blob/master/pkg/kvstore/allocator/allocator.go#L460-L463
```
			success, err := kvstore.CreateOnlyIfLocked(ctx, keyPath, []byte(k), false, lock)
			if err != nil || !success {
				return 0, false, fmt.Errorf("unable to create master key '%s': %s", keyPath, err)
			}
```